src/styles/MuiThemeProvider.js

```diff
@@ -1,20 +1,29 @@
 import {Component, PropTypes} from 'react';
 import getMuiTheme from './getMuiTheme';
+import getUniqueIdGenerator from '../utils/getUniqueIdGenerator';
+
 
 class MuiThemeProvider extends Component {
 
   static propTypes = {
     children: PropTypes.element,
     muiTheme: PropTypes.object,
+    uniqueIdGen: PropTypes.func,
   };
 
   static childContextTypes = {
     muiTheme: PropTypes.object.isRequired,
+    uniqueIdGen: PropTypes.func.isRequired,
+  };
+
+  state = {
+    uniqueIdGen: this.props.uniqueIdGen || getUniqueIdGenerator(),
```

Comment:
In your version, components in `MuiThemeProvider` will invoke all lifecycle methods(including `componentWillMount`, where `TextField` get its `uniqueId`) each time the button is clicked because the virtual DOM trees are totally different.